Speed up InverseCancellation when there's nothing to cancel

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

In the cases when there is nothing to cancel in the DAGCircuit the InverseCancellation pass would iterate over the entire circuit twice, once to search for self inverse gates and then a second time to search for other inverse pairs. This is typically fairly fast because the actual search is done in rust, but there is a python callback function that is called for each node. Depending on the size of the circuit this could add up to a significant amount of time. This commit updates the logic in the pass to first check if there are any operations being asked to cancel for either stage, and if there are then only do the iteration if there are any gates in the circuit in the set of instructions that will be cancelled, which means we'll need to do an iteration. These checks are all O(1) for any sized dag, so they're much lower overhead and will mean the pass executes much faster in the case when there isn't anything to cancel.

### Details and comments

mtreinish committed:
Speed up InverseCancellation when there's nothing to cancel

In the cases when there is nothing to cancel in the DAGCircuit the
InverseCancellation pass would iterate over the entire circuit twice,
once to search for self inverse gates and then a second time to search
for other inverse pairs. This is typically fairly fast because the
actual search is done in rust, but there is a python callback function
that is called for each node. Depending on the size of the circuit this
could add up to a significant amount of time. This commit updates the
logic in the pass to first check if there are any operations being asked
to cancel for either stage, and if there are then only do the iteration
if there are any gates in the circuit in the set of intructions that
will be cancelled, which means we'll need to do an iteration. These
checks are all O(1) for any sized dag, so they're much lower overhead
and will mean the pass executes much faster in the case when there isn't
anything to cancel.

mtreinish committed:
Speed-up when there is a partial match

Building off of the previous commit this speeds up the inverse
cancellation pass when only some of the inverse pairs are not present in
the DAG, but others are present. In this case the previous commit would
still iterate over the full dag multiple times even when we know some of
the inverse pairs are not present in the DAG. This commit updates the
logic to not call collect_runs() if we know it's going to be empty or
there is no cancellation opportunity.

mtreinish committed:
Expand test coverage

mtreinish committed:
Add more tests

jakelishman commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -76,11 +80,13 @@ def run(self, dag: DAGCircuit):
+        return dag
 
-        dag = self._run_on_self_inverse(dag, self.self_inverse_gates)
-        return self._run_on_inverse_pairs(dag, self.inverse_gate_pairs)
```
Given the immediate short-circuiting within the two functions, I doubt much is being bought by these outer `if` statements, but that's not important.

jakelishman commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -58,12 +58,16 @@ def __init__(self, gates_to_cancel: List[Union[Gate, Tuple[Gate, Gate]]]):
 
         self.self_inverse_gates = []
         self.inverse_gate_pairs = []
+        self.self_inverse_gate_names = set()
+        self.inverse_gate_pairs_names = set()
 
```
Do we actually need the gate instances?  As far as I could see, we only ever inspect the names.

mtreinish replied:
Oh that's a good point, I just assumed we needed the instances and just stored the names separately. But you're right nothing ever uses the gate instances outside of `__init__` to check things are actually inverses of each other.

mtreinish replied:
I looked into this and we do need the gate instances so that we can handle parameterized gates. The inverse pairs function uses this now, the self inverse path doesn't but this is a bug and I'm going to fix it.

jakelishman commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -91,9 +97,16 @@ def _run_on_self_inverse(self, dag: DAGCircuit, self_inverse_gates: List[Gate]):
         """
+        op_counts = dag.count_ops()
+        if not self.self_inverse_gate_names.intersection(op_counts):
+            return dag
         # Sets of gate runs by name, for instance: [{(H 0, H 0), (H 1, H 1)}, {(X 0, X 0}]
-        gate_runs_sets = [dag.collect_runs([gate.name]) for gate in self_inverse_gates]
-        for gate_runs in gate_runs_sets:
+        for gate in self.self_inverse_gates:
+            gate_name = gate.name
+            gate_count = op_counts.get(gate_name, 0)
+            if gate_count <= 1:
+                continue
+            gate_runs = dag.collect_runs([gate_name])
```
Given we calculate an intersection with `op_counts` at the top of the function, it feels like we could skip the `continue` logic by iterating over the intersection rather than manually redoing it with retrieving the counts here - that'd also remove the early return, since the for loop would already serve that purpose.

mtreinish replied:
I'm not sure we should in this case. I was thinking about it, but iterating over the intersection directly means the loop is iterating over a set and the execution order is potentially non-deterministic. I think this would be problematic from a reproducibility PoV.

jakelishman replied:
We could always sort the set - it's rather unlikely to be large enough to matter.

jakelishman commented on qiskit/transpiler/passes/optimization/inverse_cancellation.py:
```diff
@@ -123,8 +136,16 @@ def _run_on_inverse_pairs(self, dag: DAGCircuit, inverse_gate_pairs: List[Tuple[
         """
-        for pair in inverse_gate_pairs:
-            gate_cancel_runs = dag.collect_runs([pair[0].name, pair[1].name])
+        op_counts = dag.count_ops()
+        if not self.inverse_gate_pairs_names.intersection(op_counts):
+            return dag
+
+        for pair in self.inverse_gate_pairs:
+            gate_0_name = pair[0].name
+            gate_1_name = pair[1].name
+            if gate_0_name not in op_counts or gate_1_name not in op_counts:
+                continue
```
Similar comment.

jakelishman reviewed (commented):
This looks like a generally sensible improvement to me.

jakelishman commented on test/python/transpiler/test_inverse_cancellation.py:
```diff
@@ -271,6 +271,74 @@ def test_cx_do_not_wrongly_cancel(self):
+        new_circ = inverse_pass(qc)
+        self.assertEqual(qc, new_circ)
+
+    def test_some_cancel_rules_to_cancel(self):
+        """Test when there are some gates to cancel."""
+        qc = QuantumCircuit(2)
+        qc.cx(0, 1)
+        qc.cx(1, 0)
+        qc.h(0)
+        qc.h(0)
+        inverse_pass = InverseCancellation([HGate(), CXGate(), CZGate()])
+        new_circ = inverse_pass(qc)
+        self.assertNotIn("h", new_circ.count_ops())
```
Another thought: it'd be good to include a test that `cx(0, 1) / `cx(0, 1, ctrl_state=0)` doesn't cancel.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11211/commits/f2f8be2ae2ca3fc7fb444132fa1f26ae666a9386

jakelishman commented on test/python/transpiler/test_inverse_cancellation.py:
```diff
@@ -271,6 +271,74 @@ def test_cx_do_not_wrongly_cancel(self):
+        new_circ = inverse_pass(qc)
+        self.assertEqual(qc, new_circ)
+
+    def test_some_inverse_pairs(self):
+        """Test when there are some but not all inverse pairs to cancel."""
+        qc = QuantumCircuit(1)
+        qc.s(0)
+        qc.sdg(0)
+        qc.t(0)
+        qc.tdg(0)
+        inverse_pass = InverseCancellation([(TGate(), TdgGate())])
+        new_circ = inverse_pass(qc)
+        self.assertNotIn("t", new_circ.count_ops())
+        self.assertNotIn("tdg", new_circ.count_ops())
```
Another another thought: it'd be good to include a test where the inverse pair are in the circuit back-to-front, e.g. `InverseCancellation([TGate(), TdgGate()]` on a circuit `tdg(0); t(0)`.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/11211/commits/f2f8be2ae2ca3fc7fb444132fa1f26ae666a9386

mtreinish committed:
Fix handling of parameterized gates

This commit adds a fix for an issue that was caught while tuning the
performance of the pass. If a parameterized self inverse was passed in
the pass would incorrectly treat all instances of that parameterized'
gate as being a self inverse without checking the parameter values.
This commit corrects this oversight to handle this case to only cancel
gates when the optimization is correct.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into check-for-gates-inverse-cancellation

jakelishman reviewed (approved):
This looks fine, thanks.  It feels like the logic in the runners could be a shade simpler than it is, but we don't need to expand this PR to include ever more refactorings.
